contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -0,0 +1,600 @@
+#include <sys/zfs_context.h>
+#include <sys/fs/zfs.h>
+#include <sys/dsl_crypt.h>
+#include <sys/byteorder.h>
+#include <libzfs.h>
+
+#include <syslog.h>
+
+#include <sys/zio_crypt.h>
+#include <openssl/evp.h>
+
+#define	PAM_SM_AUTH
+#define	PAM_SM_PASSWORD
+#define	PAM_SM_SESSION
+#include <security/pam_modules.h>
+#include <security/pam_ext.h>
+#include <string.h>
+#include <signal.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/file.h>
+#include <sys/wait.h>
+#include <pwd.h>
+
+#include <sys/mman.h>
+static void destroy_pw(pam_handle_t *pamh, void *data, int errcode);
+typedef struct {
+	size_t len;
+	char *value;
+} password;
+
+static password *alloc_pw_size(size_t len) {
+	password *pw = malloc(sizeof (password));
+	pw->len = len;
+	pw->value = malloc(len);
+	mlock(pw->value, pw->len);
+	return (pw);
+}
+static password *alloc_pw_string(const char *source) {
+	password *pw = malloc(sizeof (password));
+	pw->len = strlen(source) + 1;
+	pw->value = malloc(pw->len);
+	mlock(pw->value, pw->len);
+	memcpy(pw->value, source, pw->len);
+	return (pw);
+}
+static void pw_free(password *pw) {
+	bzero(pw->value, pw->len);
+	munlock(pw->value, pw->len);
+	free(pw->value);
+	free(pw);
+}
+static const char *PASSWORD_VAR_NAME = "pam_zfs_key_authtok";
+static password *pw_fetch(pam_handle_t *pamh) {
+	const char *token;
+	if (pam_get_authtok(pamh, PAM_AUTHTOK, &token, NULL) != PAM_SUCCESS) {
+		pam_syslog(pamh, LOG_ERR,
+			"couldn't get password from PAM stack");
+		return (NULL);
+	}
+	return (alloc_pw_string(token));
+}
+static const password *pw_fetch_lazy(pam_handle_t *pamh) {
+	password *pw = pw_fetch(pamh);
+	if (pw == NULL) {
+		return (NULL);
+	}
+	int ret = pam_set_data(pamh, PASSWORD_VAR_NAME, pw, destroy_pw);
+	if (ret != PAM_SUCCESS) {
+		pw_free(pw);
+		pam_syslog(pamh, LOG_ERR, "pam_set_data failed");
+		return (NULL);
+	}
+	return (pw);
+}
+static const password *pw_get(pam_handle_t *pamh) {
+	const password *authtok = NULL;
+	int ret = pam_get_data(pamh, PASSWORD_VAR_NAME,
+			(const void**)(&authtok));
+	if (ret == PAM_SUCCESS)
+		return (authtok);
+	if (ret == PAM_NO_MODULE_DATA)
+		return (pw_fetch_lazy(pamh));
+	pam_syslog(pamh, LOG_ERR, "password not available");
+	return (NULL);
+}
+static int pw_clear(pam_handle_t *pamh) {
+	int ret = pam_set_data(pamh, PASSWORD_VAR_NAME, NULL, NULL);
+	if (ret != PAM_SUCCESS) {
+		pam_syslog(pamh, LOG_ERR, "clearing password failed");
+		return (-1);
+	}
+	return (0);
+}
+
+static void destroy_pw(pam_handle_t *pamh, void *data, int errcode) {
+	if (data != NULL) {
+		pw_free((password*) data);
+	}
+}
+static libzfs_handle_t *libzfs;
```

Comment:
style: this global variable should be moved to the top of the file, after the `#includes`.